### PR TITLE
729 Stock line 3

### PR DESCRIPTION
**Pull request by roxy-dao**
Another part of #729 

Allow stock line updates for: location, cost price, sell price, expiry date, batch, on hold

Will need StockLineMutate permission if testing manually... not sure which permission this is on mSupply 👀

### CODE REVIEW

File: server/graphql/stock_line/src/mutations/update.rs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        // Structured Errors
+        ServiceError::StockDoesNotExist => {
+            return Ok(UpdateErrorInterface::RecordNotFound(RecordNotFound {}))
+        }
+        ServiceError::StockIsOnHold => {
+            return Ok(UpdateErrorInterface::StockIsOnHold(StockIsOnHold {}))
+        }
```
Comment (mark-prins) [inline server/graphql/stock_line/src/mutations/update.rs:115]: can be removed I think

File: server/graphql/stock_line/src/mutations/update.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        // StockIsOnHold
+        let test_service = TestService(Box::new(|_| Err(ServiceError::StockIsOnHold)));
+        let expected = json!({
+            "updateStockLine": {
+              "error": {
+                "__typename": "StockIsOnHold"
+              }
+            }
+          }
+        );
```
Comment (mark-prins) [inline server/graphql/stock_line/src/mutations/update.rs:229]: not required?

File: server/service/src/stock_line/tests/update.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                &context,
+                inline_init(|r: &mut UpdateStockLine| {
+                    r.id = mock_stock_line_a().id;
+                    r.location_id = Some("invalid".to_string());
+                })
+            ),
+            Err(ServiceError::LocationDoesNotExist)
+        );
+
+        // StockIsOnHold
```
Comment (mark-prins) [inline server/service/src/stock_line/tests/update.rs:51]: as noted elsewhere..

File: server/service/src/stock_line/update.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    input: &UpdateStockLine,
+) -> Result<StockLineRow, UpdateStockLineError> {
+    use UpdateStockLineError::*;
+
+    let stock_line = check_stock_line_exists(connection, &input.id)?.ok_or(StockDoesNotExist)?;
+
+    if !check_store(&stock_line, store_id) {
+        return Err(StockDoesNotBelongToStore);
+    };
+    if check_stock_line_on_hold(&stock_line) {
```
Comment (mark-prins) [inline server/service/src/stock_line/update.rs:71]: and again..

File: server/service/src/stock_line/update.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        id: _,
+        location_id,
+        cost_price_per_pack,
+        sell_price_per_pack,
+        expiry_date,
+        batch,
+        on_hold,
+    }: UpdateStockLine,
+) -> StockLineRow {
+    existing.location_id = location_id.or(existing.location_id);
```
Comment (mark-prins) [inline server/service/src/stock_line/update.rs:95]: could you also add in logging please? Maybe that should be in a new PR though?

There is this sort of thing in the mSupply implementation:
```
	// 130328 v3.12 Sudhir #5209 : Log create, added source table and record id 
	If (Old([item_line]cost_price)#[item_line]cost_price)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify cost price: "+[item]code+". Old "+String(Old([item_line]cost_price))+" New: "+String([item_line]cost_price))
		log_create("price change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]sell_price)#[item_line]sell_price)
		RELATE ONE([item_line]item_ID)
		obStockLineForm.colombia.stockItemID:=[item]ID
		$s_event:=("Modify sell price: "+[item]code+". Old "+String(Old([item_line]sell_price))+" New: "+String([item_line]sell_price))
		log_create("price change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]quantity)#[item_line]quantity)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify quantity: "+[item]code+". Old "+String(Old([item_line]quantity))+" New: "+String([item_line]quantity))
		log_create("quantity change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]available)#[item_line]available)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify available: "+[item]code+". Old "+String(Old([item_line]available))+" New: "+String([item_line]available))
		log_create("quantity change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]batch)#[item_line]batch)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify batch: "+[item]code+". Old "+Old([item_line]batch)+" New: "+[item_line]batch)
		log_create("batch change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
```

and

```
		log_create("Batch off hold"; "Item line #"+[item_line]ID+" Item "+[item]code; [item_line]ID; ```
```
		log_create("Batch set to hold"; "Item line #"+[item_line]ID+" Item "+[item]code; [item_line]ID; Table(->[item_line]))  // Always log putting a stock line on hold
```

**Review by mark-prins — APPROVED** [review]
🎉  awesome! and thanks for fixing up the query too.

I have a few suggestions.. and at some point we'd need logging of changes, and querying the logs. but maybe that in another PR? this one is large enough and it would be great to get it merged so that I can add in the f/e mutation calls 😁

File: server/service/src/stock_line/update.rs (hunk 5 above)
Comment (roxy-dao) [reply server/service/src/stock_line/update.rs:95]: Do you want all the individual field changes logged? Or just have one log stating that the stock line has been updated? 🤔

Comment (mark-prins) [reply server/service/src/stock_line/update.rs:89]: individual field changes.. if a field has changed. It's a bit clumsier to code and makes more log entries.. but I believe that mSupply has done it that way for a reason and the changing of stock line information is a bit sensitive, so would prefer to be cautious

**Pull request merged**